<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.12.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scGenePT: scGenePT: Is language all you need for modeling single-cell perturbations?</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scGenePT
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.12.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){ initResizable(false); });
/* @license-end */
</script>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div id="doc-content">
<div><div class="header">
  <div class="headertitle"><div class="title">scGenePT: Is language all you need for modeling single-cell perturbations?</div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="autotoc_md0"></a></p>
<h1><a class="anchor" id="autotoc_md1"></a>
Model Description</h1>
<p>scGenePT is a collection of single-cell models for perturbation prediction. It leverages the <a href="https://github.com/bowang-lab/scGPT">scGPT</a> [1] foundation model for scRNAseq data by injecting language embeddings at the gene level into the model architecture. The language gene embeddings are obtained by embedding gene level information from different knowledge sources using LLMs. The knowledge sources used include NCBI gene descriptions, UniProt protein Summaries for protein coding genes - as inspired by the <a href="https://github.com/yiqunchen/GenePT">genePT</a> [2] approach - and GO (Gene Ontology) Gene Molecular Annotations, across three different axes: Molecular Function, Biological Process and Cellular Component</p>
<p><img src="FOSB_gene_example.png" alt="Example of gene representations for FOSB gene" class="inline"/></p>
<h1><a class="anchor" id="autotoc_md2"></a>
Training</h1>
<h2><a class="anchor" id="autotoc_md3"></a>
1. Training <br  />
</h2>
<p>scGenePT uses a pre-trained scGPT model and pre-computed gene embeddings. The files need to be under <code>models/pretrained/</code>, as described below. All of the pretrained models, as well as the assumed folder structure for model training can be found under this <b><a href="https://drive.google.com/drive/folders/1mit6pwRaykC28WQOSdPP-gFNbzKKJE2_?usp=drive_link">Google Drive Link</a></b>.</p>
<p><b>Download scGPT Pretrained Model</b> <br  />
 </p><table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Pretrained model   </th><th class="markdownTableHeadNone">Download link   </th><th class="markdownTableHeadNone">Should be under    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">scGPT Model weights (whole-human)   </td><td class="markdownTableBodyNone"><a href="https://drive.google.com/drive/folders/1oWh_-ZRdhtoGQ2Fw24HP41FgLoomVo-y">scGPT Google Drive Link</a> <br  />
 <a href="https://drive.google.com/drive/folders/1lNvQQpmHDSizEzbnKtBEheXT48OawqMc?usp=drive_link">CZI-Hosted Link</a>   </td><td class="markdownTableBodyNone"><code>models/pretrained/scgpt/</code> <br  />
 - best_model.pt <br  />
 - args.json <br  />
 - vocab.json   </td></tr>
</table>
<p><b>Download Gene Embeddings</b> <br  />
 scGenePT can use multiple sources for textual gene annotations. The different sources and gene representations are described below. For GO Gene Annotations, only one embedding choice should be used.</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Gene Embedding   </th><th class="markdownTableHeadNone">Download link   </th><th class="markdownTableHeadNone">Should be under    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">NCBI Gene summaries   </td><td class="markdownTableBodyNone"><a href="https://zenodo.org/records/10833191">GenePT zenodo Link</a> <br  />
 <a href="https://drive.google.com/file/d/1wx-CFeqp5xFdynJrmUX6a73xxxfdZ0U4/view?usp=drive_link">CZI-Hosted Link</a>   </td><td class="markdownTableBodyNone"><code>models/pretrained/genept/</code> <br  />
 NCBI_gene_embedding_ada.pickle    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">NCBI Gene summaries + UniProt protein summaries   </td><td class="markdownTableBodyNone"><a href="https://drive.google.com/file/d/1EyuQwY8B3DU3W2VBuiBSoJiJw7KHntu4/view?usp=drive_link">CZI-Hosted Link</a>   </td><td class="markdownTableBodyNone"><code>models/pretrained/gene_annotations</code> <br  />
 NCBI+UniProt_embedding_ada.pkl    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">GO Cellular Components Annotations   </td><td class="markdownTableBodyNone"><a href="https://drive.google.com/file/d/1oGnxs56GqGQA5gaocg4uPf_UwtomJ0Tp/view?usp=drive_link">CZI-Hosted Link</a>   </td><td class="markdownTableBodyNone"><code>models/pretrained/gene_annotations</code> <br  />
 GO_C_gene_embeddings_concat.pickle <b>or</b> GO_C_gene_embeddings_avg.pickle    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">GO Molecular Function Annotations   </td><td class="markdownTableBodyNone"><a href="https://drive.google.com/file/d/1ZGhHabXSg6eGkSCCMKf6o2-HvHIKpp3s/view?usp=drive_link">CZI-Hosted Link</a>   </td><td class="markdownTableBodyNone"><code>models/pretrained/gene_annotations</code> <br  />
 GO_F_gene_embeddings_concat.pickle <b>or</b> GO_F_gene_embeddings_avg.pickle    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">GO Biological Processes Annotations   </td><td class="markdownTableBodyNone"><a href="https://drive.google.com/file/d/1pVRUpth4U8zhi1mRUgF5-lNMOg4jM9FF/view?usp=drive_link">CZI-Hosted Link</a>   </td><td class="markdownTableBodyNone"><code>models/pretrained/gene_annotations</code> <br  />
 GO_P_gene_embeddings_concat.pickle <b>or</b> GO_P_gene_embeddings_avg.pickle    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Aggregation of GO-C + GO-F + GO-P   </td><td class="markdownTableBodyNone"><a href="https://drive.google.com/file/d/1cQi6CtOEESXX9iVokwlcf_onVD3OmWNk/view?usp=drive_link">CZI-Hosted Link</a>   </td><td class="markdownTableBodyNone"><code>models/pretrained/gene_annotations</code> <br  />
 GO_all_gene_embeddings_concat.pickle <b>or</b> GO_all_gene_embeddings_avg.pickle   </td></tr>
</table>
<p><b>Training Data</b> <br  />
</p>
<p>We use the processed versions of the Adamson and Norman datasets from <a href="https://github.com/snap-stanford/GEARS">GEARS</a>. Note that there are some differences in dataloaders between differrent versions, we trained and evaluated the models on GEARS v=0.0.2. This code snippet is already embedded in the the codebase. To get the training, val and test splits: </p><div class="fragment"><div class="line"><span class="keyword">from</span> GEARS <span class="keyword">import</span> PertData</div>
<div class="line">dataset_name = <span class="stringliteral">&#39;norman&#39;</span> <span class="comment"># or &#39;adamson&#39;</span></div>
<div class="line">pert_data = PertData(<span class="stringliteral">&quot;data/&quot;</span>)</div>
<div class="line">pert_data.load(data_name=dataset_name)</div>
<div class="line">pert_data.prepare_split(split=split, seed=1)</div>
<div class="line">pert_data.get_dataloader(batch_size=batch_size, test_batch_size=val_batch_size)</div>
</div><!-- fragment --><p><b>Training Script</b> <br  />
 Once the pre-trained scGPT model and pre-computed gene embeddings have been downloaded under <code>models/pretrained/</code>, as described above, scGenePT can be trained using:</p>
<p><code>python train.py --model-type=scgenept_ncbi+uniprot_gpt --num-epochs=20 --dataset=norman --device=cuda:0 --rnd-seed=42</code></p>
<p>The training arguments are the following:</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">argument   </th><th class="markdownTableHeadNone">definition   </th><th class="markdownTableHeadNone">default value    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><b>model-type</b>   </td><td class="markdownTableBodyNone">type of model to train   </td><td class="markdownTableBodyNone">scgenept_ncbi_gpt    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><b>num-epochs</b>   </td><td class="markdownTableBodyNone">number of epochs to train the model for   </td><td class="markdownTableBodyNone">20    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><b>batch-size</b>   </td><td class="markdownTableBodyNone">train batch_size   </td><td class="markdownTableBodyNone">64    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><b>eval-batch-size</b>   </td><td class="markdownTableBodyNone">validation batch_size   </td><td class="markdownTableBodyNone">64    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><b>device</b>   </td><td class="markdownTableBodyNone">device used for training   </td><td class="markdownTableBodyNone">cude:0    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><b>dataset</b>   </td><td class="markdownTableBodyNone">dataset used for training   </td><td class="markdownTableBodyNone">norman    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><b>rnd-seed</b>   </td><td class="markdownTableBodyNone">random seed used for training   </td><td class="markdownTableBodyNone">42    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><b>max-seq-len</b>   </td><td class="markdownTableBodyNone">number of genes to sample during training   </td><td class="markdownTableBodyNone">1536    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><b>dropout</b>   </td><td class="markdownTableBodyNone">dropout value to use during training   </td><td class="markdownTableBodyNone">0.2    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><b>lr</b>   </td><td class="markdownTableBodyNone">learning rate   </td><td class="markdownTableBodyNone">1e-4    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><b>schedule-interval-lr</b>   </td><td class="markdownTableBodyNone">schedule interval for lr   </td><td class="markdownTableBodyNone">1    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><b>early-stop</b>   </td><td class="markdownTableBodyNone">number of epochs to stop early if loss does not decrease   </td><td class="markdownTableBodyNone">10    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><b>log-interval</b>   </td><td class="markdownTableBodyNone">number of timesteps for which to log the loss   </td><td class="markdownTableBodyNone">100    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone"><b>pretrained-model-dir</b>   </td><td class="markdownTableBodyNone">directory of where the saved pretrained model directories are in   </td><td class="markdownTableBodyNone">models/pretrained/    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone"><b>outputs-dir</b>   </td><td class="markdownTableBodyNone">directory where model outputs and metrics are saved   </td><td class="markdownTableBodyNone">outputs/   </td></tr>
</table>
<h1><a class="anchor" id="autotoc_md4"></a>
Tutorials</h1>
<ul>
<li><a href="https://colab.research.google.com/drive/12Lg_dNy55-ii69hsfc3_bLJeVS1eNsDB">scgenept_tutorial</a> - Notebook showcasing how to use trained scGenePT models in inference mode for perturbation prediction. It uses models fine-tuned on the Norman dataset and offers examples of predicting post-perturbation expression responses for single and two-gene perturbations.</li>
</ul>
<h1><a class="anchor" id="autotoc_md5"></a>
Cite Us</h1>
<p>If you use scGenePT in your analyses, please cite us:</p>
<p><b>Paper</b>: Istrate, Ana-Maria, Donghui Li, and Theofanis Karaletsos. "scGenePT: Is language all you need for modeling single-cell perturbations?." bioRxiv (2024): 2024-10. <a href="https://www.biorxiv.org/content/10.1101/2024.10.23.619972v1">bioRxiv Link</a></p>
<div class="fragment"><div class="line">@article{istrate2024scgenept,</div>
<div class="line">  title={scGenePT: Is language all you need for modeling single-cell perturbations?},</div>
<div class="line">  author={Istrate, Ana-Maria and Li, Donghui and Karaletsos, Theofanis},</div>
<div class="line">  journal={bioRxiv},</div>
<div class="line">  pages={2024--10},</div>
<div class="line">  year={2024},</div>
<div class="line">  publisher={Cold Spring Harbor Laboratory}</div>
<div class="line">}</div>
</div><!-- fragment --><h1><a class="anchor" id="autotoc_md6"></a>
References</h1>
<ol type="1">
<li>Cui, Haotian, et al. "scGPT: toward building a foundation model for single-cell multi-omics using generative AI." Nature Methods (2024): 1-11. <a href="https://www.nature.com/articles/s41592-024-02201-0">Paper Link</a> | <a href="https://github.com/bowang-lab/scGPT">GitHub Repo</a></li>
<li>Chen, Yiqun, and James Zou. "GenePT: a simple but effective foundation model for genes and cells built from ChatGPT." bioRxiv (2024): 2023-10. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC10614824/">Paper Link</a> | <a href="https://github.com/yiqunchen/GenePT">GitHub Repo</a></li>
<li>Roohani, Yusuf, Kexin Huang, and Jure Leskovec. "Predicting transcriptional outcomes of novel multigene perturbations with GEARS." Nature Biotechnology 42.6 (2024): 927-935. <a href="https://www.nature.com/articles/s41587-023-01905-6">Paper Link</a> | <a href="https://github.com/snap-stanford/GEARS">GitHub Repo</a> </li>
</ol>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.12.0
</small></address>
</div><!-- doc-content -->
</body>
</html>
